File: Android/SayHi/app/src/main/java/de/oklemenz/sayhi/service/IconService.java
package de.oklemenz.sayhi.service;

import android.content.ContentValues;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.DisplayMetrics;

import org.jdeferred.DoneCallback;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import de.oklemenz.sayhi.AppDelegate;

import static de.oklemenz.sayhi.AppDelegate.IconsFetchedNotification;
import static de.oklemenz.sayhi.AppDelegate.Namespace;
import static de.oklemenz.sayhi.AppDelegate.SpaceSwitchedNotification;

/**
 * Created by Oliver Klemenz on 21.02.17.
 */

public class IconService extends SQLiteOpenHelper {

    private static String IconFetchDateField = Namespace + ".IconFetchDate";

    private static final int DatabaseVersion = 1;
    private static final String DatabaseName = "icons.db";

    private static IconService instance = new IconService();

    public static IconService getInstance() {
        return instance;
    }

    private static final String DbIcon = "icon";
    private static final String DbKey = "key";
    private static final String DbData = "data";

    private Map<String, Bitmap> icons = new HashMap<>();

    private IconService() {
        super(AppDelegate.getInstance().Context, DatabaseName, null, DatabaseVersion);
        NotificationCenter.getInstance().addObserver(SpaceSwitchedNotification, new NotificationCenter.Observer() {
            @Override
            public void notify(String name, NotificationCenter.Notification notification) {
                clear();
                fetch();
            }
        }, false);
    }

    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE " + DbIcon + " (" +
                DbKey + " TEXT PRIMARY KEY," +
                DbData + " TEXT)"
        );
    }

    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
    }

    public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
    }

    public void clear() {
        SharedPreferences preferences = AppDelegate.getInstance().Context.getSharedPreferences(AppDelegate.Namespace, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = preferences.edit();
        editor.putString(IconFetchDateField, null);
        editor.commit();
    }

    public void reset() {
        clear();
        IconService.this.getWritableDatabase().delete(DbIcon, null, null);
        fetch();
    }

    public void fetch() {
        final SharedPreferences preferences = AppDelegate.getInstance().Context.getSharedPreferences(AppDelegate.Namespace, Context.MODE_PRIVATE);
        Date date = null;
        String isoDate = preferences.getString(IconFetchDateField, null);
        if (isoDate != null) {
            date = Utilities.parseISO8601String(isoDate);
        }
        DataService.getInstance().fetchIcons(date).then(new DoneCallback<Map<String, String>>() {
            @Override
            public void onDone(Map<String, String> icons) {
                if (icons.size() > 0) {
                    for (Map.Entry<String, String> entry : icons.entrySet()) {
                        ContentValues values = new ContentValues();
                        values.put(DbKey, entry.getKey());
                        values.put(DbData, entry.getValue());
                        IconService.this.getWritableDatabase().replace(DbIcon, null, values);
                    }

                    Date date = new Date(System.currentTimeMillis() - 2 * 60 * 60 * 1000); // 2 hour adjustment, e.g. daylight saving time (DST)
                    String isoDate = Utilities.getISO8601StringForDate(date);
                    SharedPreferences.Editor editor = preferences.edit();
                    editor.putString(IconFetchDateField, isoDate);
                    editor.commit();
                    NotificationCenter.getInstance().post(IconsFetchedNotification);
                }
            }
        });
    }

    public Bitmap icon(String name) {
        int id = Utilities.getDrawableResourceId(AppDelegate.getInstance().Context, name);
        if (id > 0) {
            return BitmapFactory.decodeResource(AppDelegate.getInstance().Context.getResources(), id);
        }
        DisplayMetrics metrics = AppDelegate.getInstance().Context.getResources().getDisplayMetrics();
        if (metrics.density > DisplayMetrics.DENSITY_DEFAULT) {
            name = name + "@2x";
        }
        if (icons.containsKey(name)) {
            return icons.get(name);
        }
        Cursor cursor = getReadableDatabase().query(DbIcon, new String[]{DbData}, DbKey + " = ?", new String[]{name}, null, null, null);
        while (cursor.moveToNext()) {
            String data = cursor.getString(cursor.getColumnIndexOrThrow(DbData));
            byte[] bytes = Utilities.fromBase64(data);
            Bitmap icon = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
            icons.put(name, icon);
            return icon;
        }
        if (!cursor.isClosed()) {
            cursor.close();
        }
        return null;
    }
}

File: iOS/SayHi/SayHi/IconService.swift
//
//  IconService.swift
//  SayHi
//
//  Created by Klemenz, Oliver on 09.01.17.
//  Copyright © 2017 Klemenz, Oliver. All rights reserved.
//

import Foundation
import SQLite

let IconFetchDateField = "\(Namespace).IconFetchDate"

class IconService {
    
    static let instance = IconService()
    
    var db : Connection!
    let dbIcon = Table("icon")
    let dbKey = Expression<String>("key")
    let dbData = Expression<String>("data")
    
    var icons : [String:UIImage] = [:]
    
    init() {
        do {
            let documentsUrl = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
            let dbPath = documentsUrl.appendingPathComponent("icon.sqlite3")
            db = try Connection(dbPath.absoluteString)
            
            try db.run(dbIcon.create(ifNotExists: true) { t in
                t.column(dbKey, primaryKey: true)
                t.column(dbData)
            })
        } catch let error {
            print(error)
        }
        NotificationCenter.default.addObserver(self, selector: #selector(spaceSwitched), name: SpaceSwitchedNotification, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc func spaceSwitched() {
        clear()
        fetch()
    }
    
    func clear() {
        UserDefaults.standard.removeObject(forKey: IconFetchDateField)
        UserDefaults.standard.synchronize()
    }
    
    func reset() {
        clear()
        do {
            let delete = self.dbIcon.delete()
            _ = try self.db.run(delete)
        } catch let error {
            print(error)
        }
        fetch()
    }
    
    func fetch() {
        var iconFetchDate = UserDefaults.standard.string(forKey: IconFetchDateField)?.dateFromISO
        _ = DataService.instance.fetchIcons(date: iconFetchDate).then { (icons) -> () in
            if icons.count > 0 {
                for (key, data) in icons {
                    do {
                        let upsert = self.dbIcon.insert(or: .replace, self.dbKey <- key, self.dbData <- data)
                        _ = try self.db.run(upsert)
                    } catch let error {
                        print(error)
                    }
                }
                
                iconFetchDate = Date()
                iconFetchDate?.addTimeInterval(-2*60*60) // 2 hour adjustment, e.g. daylight saving time (DST)
                UserDefaults.standard.set(iconFetchDate?.iso, forKey: IconFetchDateField)
                UserDefaults.standard.synchronize()
                
                NotificationCenter.default.post(name: IconsFetchedNotification, object: nil)
            }
        }
    }
    
    func icon(_ name: String) -> UIImage? {
        if let image = UIImage(named: name) {
            return image
        }
        var name = name
        if UIScreen.main.isRetina {
            name = "\(name)@2x"
        }
        if icons[name] != nil {
            return icons[name]
        }
        do {
            for icon in try db.prepare(dbIcon.filter(dbKey == name)) {
                if let data = Data(base64Encoded: icon[dbData]) {
                    if var image = UIImage(data: data) {
                        image = UIImage(cgImage: image.cgImage!, scale: UIScreen.main.screenScale, orientation: .up)
                        icons[name] = image
                        return image
                    }
                }
            }
        } catch let error {
            print(error)
        }
        return nil
    }
}
